tensorflow/compiler/xla/service/gpu/target_util.cc

```diff
@@ -67,6 +69,30 @@ struct TargetIntrinsics GetIntrinsic(TargetIntrinsicID intrin) {
       return {llvm::Intrinsic::nvvm_barrier0,
               llvm::Intrinsic::amdgcn_s_barrier};
     }
+    case TargetIntrinsicID::kBlockDimx: {
+      return {llvm::Intrinsic::nvvm_read_ptx_sreg_ntid_x,
+              [](llvm
+                 : IRBuilder<>* b_) -> llvm::Value* {
+                return GpuElementalIrEmitter::EmitDeviceFunctionCall(
```

Comment:
This looks like a hack to me. I digged a little bit on the llvm side and find the implementation of `nvvm_read_ptx_sreg_ntid_x` [here](https://llvm.org/docs/doxygen/NVVMIntrRange_8cpp_source.html). The implementation cached the call result to range metadata and avoids repeatedly calling into the actual assembly. It looks like we need exactly this pass. Shall we file tools team a ticket for it?

Reply:
I don't quite follow you here. My intention here is to cope with the fact that to get the number of workgroups/blocks, CUDA and ROCm has different implementation.

On CUDA, `nvvm_read_ptx_sreg_ntid_x` LLVM intrinsic is being used.
On ROCm, there is no such intrinsic in LLVM but a device function `__ockl_get_local_size` should be used.

Reply:
@jerryyin Also please refer to LLVM language ref ( https://llvm.org/docs/LangRef.html#range-metadata ). `!range` metadata provides a hint to the backend. But number of workgroups/blocks can really be determined at runtime, based on the parameters specified at kernel launch sites. So the LLVM opt pass you mentioned doesn't really affect what's being proposed here.

Reply:
@jerryyin Thanks for your pointers. Looks like I was having wrong assumptions with the llvm pass. My original question means to ask if there's anyway to add a similar intrinsic on amdgpu target, just so that it will bring the implementation to parallel?

Reply:
+ @b-sumner as he should possess more in-depth context on the rationale behind the design decision these operations are made as device functions instead of intrinsics.

@jerryyin Relevant interfaces in `ROCm-Device-Libs` can be located at:
https://github.com/RadeonOpenCompute/ROCm-Device-Libs/blob/master/ockl/inc/ockl.h#L437

My assumption is that there's no plan to add these device functions as LLVM intrinsics on AMDGPU target.

To cope with upcoming MLIR, we'll also need an MLIR opt pass to convert workgroup/workitem indexing operators in MLIR GPU Dialect to device functions in ROCm-Device-Libs, and LLVM intrinsics exposed in AMDGPU target.